experimental/incremental/task.go

```diff
@@ -548,30 +533,43 @@ func (t *task) run(caller *Task, q *AnyQuery, async bool) (output *result) {
 	}
 
 	if debugIncremental {
-		fmt.Fprintf(os.Stderr,
-			"incremental: executing: %p/%d, %T/%v\n",
-			caller.exec, caller.runID, q.Underlying(), q.Underlying())
+		callee.log("executing", "%T/%v", q.Underlying(), q.Underlying())
 	}
 
 	output.Value, output.Fatal = q.Execute(callee)
 	output.runID = callee.runID
 
 	if debugIncremental {
-		fmt.Fprintf(os.Stderr,
-			"incremental: returning: %p/%d, %T/%v\n",
-			caller.exec, caller.runID, q.Underlying(), q.Underlying())
+		callee.log("returning", "%T/%v", q.Underlying(), q.Underlying())
 	}
 
 	return output
 }
 
 // waitUntilDone waits for this task to be completed by another goroutine.
-func (t *task) waitUntilDone(caller *Task) *result {
+func (t *task) waitUntilDone(caller *Task, async bool) *result {
+	// If this task is being executed synchronously with its caller, we need to
+	// drop our semaphore hold, otherwise we will deadlock: this caller will
+	// be waiting for the leader of this task to complete, but that one
+	// may be waiting on a semaphore hold, which it will not acquire due to
+	// tasks waiting for it to complete holding the semaphore in this function.
+	//
+	// If the task is being executed asynchronously, this function is not
+	// called while the semaphore is being held, which avoids the above
+	// deadlock scenario.
+	if !async {
+		caller.release()
+	}
```

Comment:
Is there a case where a synchronous task (e.g. the root task) will call `release()` and be aborted as a bad release while waiting to become the executor/leader?

Reply:
That can happen due to concurrency bugs, but it will bubble up as a panic. There's already tests for aborts doing this correctly.